internal/db/db.go

```diff
@@ -38,37 +38,51 @@ func (db *DB) Open(bootstrap bool, project string) error {
 	newSchema := db.Schema()
 	if !bootstrap {
 		checkVersions := func(ctx context.Context, current int, tx *sql.Tx) error {
-			schemaVersion := newSchema.Version()
-			err = cluster.UpdateClusterMemberSchemaVersion(tx, schemaVersion, db.listenAddr.URL.Host)
+			internalSchemaVersion, externalSchemaVersion := newSchema.Version()
+			err = cluster.UpdateClusterMemberSchemaVersion(tx, internalSchemaVersion, externalSchemaVersion, db.listenAddr.URL.Host)
 			if err != nil {
 				return fmt.Errorf("Failed to update schema version when joining cluster: %w", err)
 			}
 
-			versions, err := cluster.GetClusterMemberSchemaVersions(ctx, tx)
+			internalVersions, externalVersions, err := cluster.GetClusterMemberSchemaVersions(ctx, tx)
 			if err != nil {
 				return fmt.Errorf("Failed to get other members' schema versions: %w", err)
 			}
 
-			for _, version := range versions {
-				if schemaVersion == version {
-					// Versions are equal, there's hope for the
-					// update. Let's check the next node.
-					continue
+			for i, versions := range [][]uint64{internalVersions, externalVersions} {
+				schemaVersion := internalSchemaVersion
+				if i == 1 {
+					schemaVersion = externalSchemaVersion
 				}
 
-				if schemaVersion > version {
-					// Our version is bigger, we should stop here
-					// and wait for other nodes to be upgraded and
-					// restarted.
-					otherNodesBehind = true
-					return schema.ErrGracefulAbort
+				for _, version := range versions {
+					if schemaVersion == version {
+						// Versions are equal, there's hope for the
+						// update. Let's check the next node.
+						continue
+					}
+
+					if schemaVersion > version {
+						// Our version is bigger, we should stop here
+						// and wait for other nodes to be upgraded and
+						// restarted.
+						otherNodesBehind = true
+						break
+					}
+
+					// Another node has a version greater than ours
+					// and presumeably is waiting for other nodes
+					// to upgrade. Let's error out and shutdown
+					// since we need a greater version.
+					otherNodesBehind = false
+					return fmt.Errorf("this node's version is behind, please upgrade")
```

Comment:
I'm finding this block a little hard to read. Would it be clearer if we define a function for the inner loop? Something like:
```golang
checkVersion := func (schemaVersion uint64, versions []uint64) error {
	for _, version := range versions {
		if schemaVersion == version {
			// Versions are equal, there's hope for the
			// update. Let's check the next node.
			continue
		}

		if schemaVersion > version {
			// Our version is bigger, we should stop here
			// and wait for other nodes to be upgraded and
			// restarted.
			otherNodesBehind = true
			break
		}

		// Another node has a version greater than ours
		// and presumeably is waiting for other nodes
		// to upgrade. Let's error out and shutdown
		// since we need a greater version.
		otherNodesBehind = false
		return fmt.Errorf("this node's version is behind, please upgrade")
	}
}
```

Then have two calls, one for the internal version and one for the external:
```golang
err = checkVersion(internalSchemaVersion, internalVersions)
if err != nil {
	return err
}

err = checkVersion(externalSchemaVersion, externalVersions)
if err != nil {
	return err
}
```

Reply:
I like this, though it would mean @gabrielmougard's PR would also need to implement a second function for comparing API extensions.

Reply:
It's fine by me :+1: